src/server/qtquick/wsurfaceitem.h

```diff
@@ -43,11 +43,13 @@ class WAYLIB_SERVER_EXPORT WSurfaceItemContent : public QQuickItem, public WObje
     friend class WSurfaceItem;
     friend class WSurfaceItemPrivate;
     friend class WSGTextureProvider;
+    friend class WSGRenderFootprintNode;
 
     void componentComplete() override;
     QSGNode *updatePaintNode(QSGNode *, UpdatePaintNodeData *) override;
     void releaseResources() override;
     void itemChange(ItemChange change, const ItemChangeData &data) override;
+    bool rendered = false;
```

Comment:
Needs QAtomicInt?

Reply:
seems QSG render begin => call fpnode's render => output buffer commit => call slot func, happens sequentially

Reply:
ok, render and qml are in different threads, maybe is safer.